### PR TITLE
Add support for custom cell renderers in Table

### PR DESCRIPTION
**Pull request by lvreynoso**
### Description

Allows custom cell renderers to be declared for a specific key on an item, which means we can easily write custom JSX for a specific column. This is to enable a modal to launch when users click on a phylo tree name.

#### Issue
[ch120777](https://app.clubhouse.io/genepi/story/120777)

### Test plan

From a UI perspective, everything should look the same and render properly. Load app, go to data tables

### CODE REVIEW

File: src/ts/src/common/components/library/data_table/index.tsx [hunk 1]
```diff
@@ -27,29 +34,30 @@ const DataTable: FunctionComponent<Props> = ({ data = [], headers }: Props) => {
...
-            }
-            if (index === 0) {
-                icon = ICONS[item.type];
+    const sampleRow = (
+        item: Record<string | number, JSONPrimitive>
+    ): Array<JSX.Element> => {
+        return headers.map((header, index) => {
+            const value = item[header.key];
+            if (renderer !== undefined) {
+                console.log("used created renderer");
```
Comment (ttung) [inline src/ts/src/common/components/library/data_table/index.tsx]: intentional?

File: src/ts/src/common/components/library/data_table/index.tsx [hunk 2]
```diff
@@ -27,29 +34,30 @@ const DataTable: FunctionComponent<Props> = ({ data = [], headers }: Props) => {
...
-                displayData = UNDEFINED_TEXT;
-            }
-            if (index === 0) {
-                icon = ICONS[item.type];
+    const sampleRow = (
+        item: Record<string | number, JSONPrimitive>
+    ): Array<JSX.Element> => {
+        return headers.map((header, index) => {
+            const value = item[header.key];
+            if (renderer !== undefined) {
```
Comment (ttung) [inline src/ts/src/common/components/library/data_table/index.tsx]: suggestion:

```
if (renderer === undefined) {
  renderer = defaultCellRenderer;
}
return (
    <Table.Cell key={`${item[indexingKey]}-${header.key}`}>
        {renderer({value, header, index})}
    </Table.Cell>
);

File: src/ts/src/common/components/library/data_subview/index.tsx [hunk 3]
```diff
@@ -7,8 +7,9 @@ import { DataTable } from "common/components";
 import style from "./index.module.scss";
 
 interface Props {
-    data?: BioinformaticsData[];
+    data?: Record<string | number, JSONPrimitive>[];
```
Comment (ttung) [inline src/ts/src/common/components/library/data_subview/index.tsx]: Why is this moving from BioinformaticsData to this new construct?

Comment (lvreynoso) [reply src/ts/src/common/components/library/data_subview/index.tsx]: To make it more generic, if we want to reuse it.

Comment (lvreynoso) [reply src/ts/src/common/components/library/data_subview/index.tsx]: `Record` is a built in Typescript type.

**Comment by phoenixAja** [comment]
we should add @tihuan as a reviewer for this PR too

File: src/ts/src/common/types/table.d.ts [hunk 4]
```diff
@@ -1,5 +1,18 @@
 interface Header {
     [index: string]: string;
     text: string;
-    key: keyof BioinformaticsType;
+    key: string | number;
```
Comment (phoenixAja) [inline src/ts/src/common/types/table.d.ts]: shouldn't this just always be a number?

Comment (lvreynoso) [reply src/ts/src/common/types/table.d.ts]: `key` here refers to the keys of a JS object, like on a sample or tree, which will almost always be strings.

File: src/ts/src/common/components/library/data_subview/index.tsx (hunk 3 above)
Comment (tihuan) [reply src/ts/src/common/components/library/data_subview/index.tsx]: SGTM!

To keep it DRY, should we declare a `Data` type and use it throughout the file?

E.g.,

```ts
// <DataTable /> src/ts/src/common/components/library/data_table/index.tsx
export type Data = Record<string | number, JSONPrimitive>;

// <DataSubview /> src/ts/src/common/components/library/data_subview/index.tsx
import { DataTable, Data } from 'src/common/components/library/data_table'

interface SearchState {
    searching?: boolean;
    results?: Data[]; // <-- this!
}
...

```

File: src/ts/src/common/components/library/data_table/index.tsx [hunk 5]
```diff
@@ -1,66 +1,73 @@
...
+    return <div className={style.cell}>{displayData}</div>;
+}
 
-  function tableRows(data: BioinformaticsData[]): Array<JSX.Element> {
-    return data.map((item) => (
-      <Table.Row key={`${item[indexingKey]}`}>{sampleRow(item)}</Table.Row>
+const DataTable: FunctionComponent<Props> = ({
+    data = [],
+    headers,
+    renderer,
```
Comment (tihuan) [inline src/ts/src/common/components/library/data_table/index.tsx]: we can add default value to renderer here, so we don't need to reassign it later!

E.g.,

```tsx
const DataTable: FunctionComponent<Props> = ({
    data = [],
    headers,
    renderer = defaultCellRenderer, // <-- this!
}: Props) => {
```

File: src/ts/src/common/components/library/data_table/index.tsx [hunk 6]
```diff
@@ -1,66 +1,73 @@
...
+            <div className={style.headerCell}>{column.text}</div>
+        </Table.HeaderCell>
     ));
-  }
+    const sampleRow = (
+        item: Record<string | number, JSONPrimitive>
+    ): Array<JSX.Element> => {
+        return headers.map((header, index) => {
+            const value = item[header.key];
+            if (renderer === undefined) {
```
Comment (tihuan) [inline src/ts/src/common/components/library/data_table/index.tsx]: we can remove this after adding default value

**Review by tihuan — COMMENTED** [review]
PR looks great! Just some super minor nits 🤩 👍 Thank you!!

File: src/ts/tsconfig.json [hunk 7]
```diff
@@ -2,7 +2,7 @@
   "compilerOptions": {
     "allowJs": true,
     "allowSyntheticDefaultImports": true,
-    "baseUrl": ".",
+    "baseUrl": "./src",
```
Comment (tihuan) [inline src/ts/tsconfig.json]: Ah sorry I changed the `baseUrl` to `"."`, so it's easier to use autocomplete in important statement starting with `import from 'src/...'` and you get a dropdown of of options 😆

File: src/ts/src/common/types/table.d.ts (hunk 4 above)
Comment (tihuan) [reply src/ts/src/common/types/table.d.ts]: Should we just restrict the type of `key` to be string?

The current use cases seem to be string anyway 😆 

```ts
export const SAMPLE_HEADERS: Header[] = [
  {
    key: "privateId",
    text: "Private ID",
  },
  {
    key: "publicId",
    text: "Public ID",
  },
  {
    key: "uploadDate",
    text: "Upload Date",
  },
  {
    key: "collectionDate",
    text: "Collection Date",
  },
  {
    key: "collectionLocation",
    text: "Collection Location",
  },
  {
    key: "gisaid",
    text: "GISAID",
  },
];

export const TREE_HEADERS: Header[] = [
  {
    key: "name",
    text: "Tree Name",
  },
  {
    key: "creationDate",
    text: "Creation Date",
  },
  {
    key: "pathogenGenomeCount",
    text: "Total Samples",
  },
  {
    key: "downloadLink",
    text: "",
  },
];

```

File: src/ts/tsconfig.json (hunk 7 above)
Comment (lvreynoso) [reply src/ts/tsconfig.json]: Ah, I use Sublime and my TS plugins are not that extensive; I've tried to organize code so we are mostly importing from the current directory or somewhere in `common` anyways.

Comment (tihuan) [reply src/ts/tsconfig.json]: Got it! What if we add another shortcut to import directly from "common"? Would that help?

```ts
    ...
    "baseUrl": ".",
    ...
    "paths": {
      "src/*": [
        "./src/*"
      ],
      "common/*": [ // <-- this!
        "./src/common/*"
      ]
    },
```

**Review by tihuan — APPROVED** [review]
LGTM! I added `src/*` back for now, so the app compiles and we can get the PR in for demo. And we can find another time to figure out the important statement strategy that works for everyone 🙏 Thank you!

**Pull request merged**